pkg/controller/ipam/antrea_ipam_controller.go

```diff
@@ -187,40 +204,124 @@ func (c *AntreaIPAMController) cleanIPPoolForStatefulSet(namespacedName string)
+func (c *AntreaIPAMController) preallocateIPPoolForStatefulSet(ss *appsv1.StatefulSet) error {
+	klog.InfoS("Processing create notification", "Namespace", ss.Namespace, "StatefulSet", ss.Name)
+
+	ipPools := c.getIPPoolsForStatefulSet(ss)
+
+	if ipPools == nil {
+		// nothing to preallocate
+		return nil
+	}
+
+	if len(ipPools) > 1 {
+		return fmt.Errorf("annotation of multiple IP Pools is not supported")
+	}
+
+	// Only one pool is supported for now. Dual stack support coming in future.
+	ipPoolName := ipPools[0]
+	allocator, err := poolallocator.NewIPPoolAllocator(ipPoolName, c.crdClient, c.ipPoolLister)
+	if err != nil {
+		return fmt.Errorf("failed to find IP Pool %s: %s", ipPoolName, err)
+	}
+
+	size := int(*ss.Spec.Replicas)
+	// Note that AllocateStatefulSet would not preallocate IPs if this StatefulSet is already present
+	// in the pool. This safeguards us from double allocation in case agent allocated IP by the time
+	// controller task is executed. Note also that StatefulSet resize will not be handled.
+	err = allocator.AllocateStatefulSet(ss.Namespace, ss.Name, size)
+	if err != nil {
+		return fmt.Errorf("failed to preallocate continuous IP space of size %d from Pool %s: %s", size, ipPoolName, err)
+	}
+
+	return nil
+}
 func (c *AntreaIPAMController) statefulSetWorker() {
 	for c.processNextStatefulSetWorkItem() {
 	}
 }
 
 func (c *AntreaIPAMController) processNextStatefulSetWorkItem() bool {
-	key, quit := c.statefulSetCleanupQueue.Get()
+	key, quit := c.statefulSetQueue.Get()
 	if quit {
 		return false
 	}
 
-	defer c.statefulSetCleanupQueue.Done(key)
-	err := c.cleanIPPoolForStatefulSet(key.(string))
+	defer c.statefulSetQueue.Done(key)
+
+	namespacedName := key.(string)
+	namespace, name := k8s.SplitNamespacedName(namespacedName)
+	ss, err := c.statefulSetInformer.Lister().StatefulSets(namespace).Get(name)
 
 	if err != nil {
-		// Put the item back on the workqueue to handle any transient errors.
-		c.statefulSetCleanupQueue.AddRateLimited(key)
-		klog.ErrorS(err, "Failed to clean IP Pool for StatefulSet", "StatefulSet", key)
-		return true
+		if errors.IsNotFound(err) {
+			// StatefulSet no longer present - clean up reserved pool IPs
+			err = c.cleanIPPoolForStatefulSet(namespacedName)
+			if err != nil {
+				// Put the item back on the workqueue to handle any transient errors.
+				c.statefulSetQueue.AddRateLimited(key)
+				klog.ErrorS(err, "Failed to clean IP Pool", "StatefulSet", key)
+				return true
+			}
+		} else {
+			klog.ErrorS(err, "Failed to get", "StatefulSet", key)
+		}
+	} else {
+		// StatefulSet was created - preallocate IPs based on replicas with best effort
+		err := c.preallocateIPPoolForStatefulSet(ss)
+		if err != nil {
+			// Preallocation is best effort - we do not retry even with transient errors,
+			// since we don't want to implement logic that would delay Pods while waiting for
+			// preallocation.
+			klog.ErrorS(err, "No IPs preallocated")
```

Comment:
```suggestion
			klog.ErrorS(err, "No IPs preallocated for StatefulSet", "StatefulSet", key)
```
Otherwise we don't even know which one fails.

Reply:
Done